### PR TITLE
LINK-1648 | 1/3 MenuDropdown component

### PR DESCRIPTION
**Pull request by jorilindell**
## Description
MenuDropdown component which is used to select action in attendance list and signups pages

This one of the 3 PRs that will implement signups page. PRs are splitted to following parts:
- **MenuDropdown component**
- Attendance list page action buttons (https://github.com/City-of-Helsinki/linkedregistrations-ui/pull/96)
- Signups page (https://github.com/City-of-Helsinki/linkedregistrations-ui/pull/97)

## Partially closes
[LINK-1648](https://helsinkisolutionoffice.atlassian.net/browse/LINK-1648)

## Screenshots
Screenshot 2023-10-26 at 19 24 10
Screenshot 2023-10-26 at 19 24 15

[LINK-1648]: https://helsinkisolutionoffice.atlassian.net/browse/LINK-1648?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/common/components/menuDropdown/__tests__/MenuDropdown.test.tsx [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    screen.getByRole('button', { name: items[index].label });
+
+  return {
+    getItemAtIndex,
+    getElement,
+  };
+};
+
+const getElement = (key: 'toggleButton') => {
+  switch (key) {
```
Comment (harriris-vincit) [inline src/common/components/menuDropdown/__tests__/MenuDropdown.test.tsx:34]: With just one value to match the `switch` statement seems unnecessary for the current implementation (although it looks like it has been used in a similar way elsewhere too).

File: src/hooks/useDropdownKeyboardNavigation.ts [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+  const handleKeyDown = React.useCallback(
+    (event: KeyboardEvent) => {
+      // Handle keyboard events only if current element is focused
+      if (!isComponentFocused()) return;
+
+      onKeyDown && onKeyDown(event);
+      switch (event.key) {
+        case 'ArrowUp':
+          if (isInitialNavigation && typeof initialFocusedIndex === 'number') {
```
Comment (harriris-vincit) [inline src/hooks/useDropdownKeyboardNavigation.ts:65]: The `ArrowUp` and `ArrowDown` cases are very similar and could be handled by a single function:
```
const handleArrowKey = (direction: 'up' | 'down', indexModifier: number = 0) => {
    const targetIndex = isInitialNavigation && typeof initialFocusedIndex === 'number'
      ? initialFocusedIndex + indexModifier
      : focusedIndex;

    focusOption(direction, targetIndex);
    event.preventDefault();
};
```
And then `handleArrowKey('up');` or `handleArrowKey('down', -1);`

File: src/common/components/menuDropdown/menu/Menu.tsx [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+  setFocusedIndex,
+  ...rest
+}) => {
+  const menuStyles: MenuStyles = React.useMemo(() => {
+    const { height, right, top, width } = menuContainerSize;
+    // the menu width should be at least 190px
+    const minWidth =
+      width < MENU_MIN_WIDTH
+        ? MENU_MIN_WIDTH
+        : /* istanbul ignore next */ width;
```
Comment (harriris-vincit) [inline src/common/components/menuDropdown/menu/Menu.tsx:53]:
```suggestion
    const minWidth = Math.max(width, MENU_MIN_WIDTH);
```

File: src/common/components/menuDropdown/MenuDropdown.tsx [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+      setMenuOpen(true);
+    }
+  };
+
+  const setFocusToButton = () => {
+    toggleButton.current?.focus();
+  };
+
+  const handleItemClick = () => {
+    if (closeOnItemClick) {
```
Comment (harriris-vincit) [inline src/common/components/menuDropdown/MenuDropdown.tsx:78]: SonarCloud is showing that these lines are uncovered by tests

File: src/common/components/menuDropdown/menu/Menu.tsx (hunk 3 above)
Comment (jorilindell) [reply src/common/components/menuDropdown/menu/Menu.tsx]: Changed

File: src/hooks/useDropdownKeyboardNavigation.ts (hunk 2 above)
Comment (jorilindell) [reply src/hooks/useDropdownKeyboardNavigation.ts]: Changed

File: src/common/components/menuDropdown/MenuDropdown.tsx (hunk 4 above)
Comment (jorilindell) [reply src/common/components/menuDropdown/MenuDropdown.tsx:78]: Added a test to cover this.

File: src/common/components/menuDropdown/__tests__/MenuDropdown.test.tsx (hunk 1 above)
Comment (jorilindell) [reply src/common/components/menuDropdown/__tests__/MenuDropdown.test.tsx]: I have started to refactor switch statements with less that 3 cases. I'll change them when they are related to the current PR. Missed this one but it's changed now.

**Comment by harriris-vincit** [comment]
@jorilindell If my user is not strongly identified and the action menu with a single disabled Merkkaa läsnäolijat button is used with up and down arrow keys (keyboard), my browser (Firefox) freezes completely and I need to stop the JavaScript execution to make the page work again

**Comment by jorilindell** [comment]
> @jorilindell Noticed a possible serious bug: if my user is not strongly identified and the action menu with a single disabled Merkkaa läsnäolijat button is used with up and down arrow keys (keyboard), my browser (Firefox) freezes completely and I need to stop the JavaScript execution to make the page work again

Good finding. It's caused by infinite while loop in `focusOption` function in `useDropdownKeyboardNavigation` hook. Fixed the issue. Still have to do PR to Linked Components UI repo, where this same issue most probably exists as well.

**Review by harriris-vincit — APPROVED** [review]
Can confirm that the freezing issue is now gone. LGTM otherwise.

**Pull request merged**